aten/src/ATen/cuda/detail/CUDAHooks.cpp

```diff
@@ -459,6 +462,134 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
   c10::cuda::device_synchronize();
 }
 
+std::tuple<size_t, size_t, ptrdiff_t, std::string, std::string, std::string, uint64_t, bool>
+CUDAHooks::StorageShareDevice(const c10::Storage& storage) const {
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+    auto shandle = c10::cuda::CUDACachingAllocator::shareIpcHandle(storage.mutable_data());
+    size_t ipc_memory_handle_size = shandle.handle.size();
+    size_t ipc_event_handle_size = sizeof(cudaIpcEventHandle_t);
+    std::string memory_handle(ipc_memory_handle_size, '\0');
+    std::string event_handle(ipc_event_handle_size, '\0');
```

Comment:
Alright, I was wondering about this, but I think I understand now. Your problem is that you need a char* and length to build the Python bytes string from. But you cannot directly return shandle, as it is a CUDA specific type. But you have another problem which is shandle dies when it exits scope here, which means you can't pass its internal char* directly. So you copy it into an owning vector so that you can keep it live.

I guess the handle size is not that long. But I still feel bad about doing an extra copy here. The other approach we could take here is to have a PIMPL we return which has a virtualized destructor which can deallocate shandle after we take care of it. You know what, this is fine.